File: 总题库/224.BasicCalculator.py
class Solution:
    def calculate(self, s: str) -> int:
        st = []
        res = 0
        sLen = len(s)
        sign = 1
        i = 0

        while i < sLen:
            num = 0
            if s[i].isdigit():
                while i < sLen and s[i].isdigit():
                    num = 10 * num + int(s[i])
                    i += 1
                i -= 1
            res += num * sign
            if s[i] == '-': sign = -1
            if s[i] == '+': sign = 1
            if s[i] == '(':
                st.append(res)
                st.append(sign)
                res, sign = 0, 1
            if s[i] == ')':
                res *= st[-1]
                st.pop()
                res += st[-1]
                st.pop()
            i += 1
        
        return res

File: 总题库/224.BasicCalculator.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> st;
        int res = 0, n = s.size(), sign = 1;
        for (int i = 0; i < n; ++i) {
            int num = 0;
            if (isdigit(s[i])) {
                while (i < n && isdigit(s[i])) {
                    num = 10 * num + (s[i] - '0');
                    i++;
                }
                i--;
                res += sign * num;
            }
            if (s[i] == '-') sign = -1;
            if (s[i] == '+') sign = 1;
            if (s[i] == '(') {
                st.push(res);
                st.push(sign);
                res = 0;
                sign = 1;
            }
            if (s[i] == ')') {
                res *= st.top();
                st.pop();
                res += st.top();
                st.pop();
            }
        }
        return res;
    }
};